File: src/G5.Logic/PreFlopParams.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;



namespace G5.Logic
{
    public struct PreFlopParams
    {
        public TableType TableType { get; private set; }
        public Position Position { get; private set; }
        public int NumCallers { get; private set; }
        public int NumRaises { get; private set; }
        public int NumActivePlayers { get; private set;  }
        public ActionType PreviousAction { get; private set; }
        public bool InPositionOnFlop { get; private set; }

        public PreFlopParams(TableType tableType, Position position, int numCallers, int numRaises, int numActivePlayers,
            ActionType previousAction, bool inPositionOnFlop) : this()
        {
            TableType = tableType;
            Position = position;
            NumCallers = numCallers;
            NumRaises = numRaises;
            NumActivePlayers = numActivePlayers;
            PreviousAction = previousAction;
            InPositionOnFlop = inPositionOnFlop;
        }

        public bool ForcedAction()
        {
            bool forced = true;

            if (Position == Position.BigBlind)
            {
                forced = NumRaises > 0;
            }

            return forced;
        }

        public static List<PreFlopParams> getAllParams(TableType tableType)
        {
            var allParams = new List<PreFlopParams>();

            if (tableType == TableType.HeadsUp)
            {
                for (int numRaises = 0; numRaises <= 4; numRaises++)
                    allParams.Add(new PreFlopParams(tableType, Position.SmallBlind, 0, numRaises, 2, ActionType.Fold, true));

                for (int numRaises = 0; numRaises <= 4; numRaises++)
                    allParams.Add(new PreFlopParams(tableType, Position.BigBlind, 0, numRaises, 2, ActionType.Fold, false));
            }
            else
            {
                foreach (Position pos in Enum.GetValues(typeof(Position)))
                {
                    allParams.Add(new PreFlopParams(tableType, pos, 0, 0, (int)tableType, ActionType.Fold, false));
                    allParams.Add(new PreFlopParams(tableType, pos, 1, 0, (int)tableType, ActionType.Fold, false));
                    allParams.Add(new PreFlopParams(tableType, pos, 0, 1, (int)tableType, ActionType.Fold, false));
                    allParams.Add(new PreFlopParams(tableType, pos, 1, 1, (int)tableType, ActionType.Fold, false));
                    allParams.Add(new PreFlopParams(tableType, pos, 0, 2, (int)tableType, ActionType.Fold, false));
                }

                ActionType[] prevActions = { ActionType.Check, ActionType.Raise };

                foreach (var previousAction in prevActions)
                {
                    for (int inPos = 0; inPos <= 1; inPos++)
                    {
                        for (int numPl = 2; numPl <= 3; numPl++)
                        {
                            allParams.Add(new PreFlopParams(tableType, Position.BigBlind, 0, 1, numPl, previousAction, inPos != 0));
                            allParams.Add(new PreFlopParams(tableType, Position.BigBlind, 1, 1, numPl, previousAction, inPos != 0));
                            allParams.Add(new PreFlopParams(tableType, Position.BigBlind, 0, 2, numPl, previousAction, inPos != 0));
                        }
                    }
                }
            }

            return allParams;
        }

        public int ToIndex()
        {
            int index;

            if (TableType == TableType.HeadsUp)
            {
                if (Position == Position.SmallBlind)
                {
                    index = Math.Min(NumRaises, 4);
                }
                else if (Position == Position.BigBlind)
                {
                    index = 5 + Math.Min(NumRaises, 4);
                }
                else
                {
                    throw new Exception("Wrong position for HU table: " + Position);
                }
            }
            else if (PreviousAction == ActionType.Fold) // Prvi krug, prvih 30 pozicija...
            {
                int a0 = (int)Position;
                int a1 = -1;

                if (NumRaises == 0)
                {
                    a1 = (NumCallers == 0) ? 0 : a1;
                    a1 = (NumCallers > 0) ? 1 : a1;
                }
                else if (NumRaises == 1)
                {
                    a1 = (NumCallers == 0) ? 2 : a1;
                    a1 = (NumCallers > 0) ? 3 : a1;
                }
                else if (NumRaises >= 2)
                {
                    a1 = 4;
                }
 
                Debug.Assert((a0 != -1) && (a1 != -1));
                index = (5 * a0) + a1; 
            }
            else // Drugi krug. PreviousAction: Check/Call, Raise/Bet
            {
                int a0 = -1;
                int a1 = -1;
                int a2 = -1;
                int a3 = -1;

                a0 = (PreviousAction == ActionType.Check || PreviousAction == ActionType.Call) ? 0 : a0;
                a0 = (PreviousAction == ActionType.Bet || PreviousAction == ActionType.Raise) ? 1 : a0;

                a1 = InPositionOnFlop ? 0 : 1;

                a2 = (NumActivePlayers == 2) ? 0 : a2;
                a2 = (NumActivePlayers > 2) ? 1 : a2;

                if (NumRaises == 1)
                {
                    a3 = (NumCallers == 0) ? 0 : a3;
                    a3 = (NumCallers > 0) ? 1 : a3;
                }
                else if (NumRaises >= 2)
                {
                    a3 = 2;
                }

                Debug.Assert((a0 != -1) && (a1 != -1) && (a2 != -1) && (a3 != -1));
                index = 30 + (12 * a0) + (6 * a1) + (3 * a2) + a3;
            }

            return index;
        }

        public override string ToString()
        {
            if (PreviousAction == ActionType.Fold)
            {
                return PreviousAction.ToString() + ", " + Position + ", bets: " + NumRaises + ", limpers: " + NumCallers;
            }
            else
            {
                string posTrs = InPositionOnFlop ? "inp" : "oop";
                return PreviousAction.ToString() + ", " + posTrs + ", players: " + NumActivePlayers + ", raises: " + NumRaises + ", callers: " + NumCallers;
            }
        }
    };
}

File: src/DecisionMaking/PreFlopParams.cpp
#include "PreFlopParams.h"
#include <assert.h>
#include <stdexcept>
#include <algorithm>


namespace G5Cpp
{
    PreFlopParams::PreFlopParams(TableType tt, Position position, int numCallers, int numRaises, int numPlayers, ActionType previousAction, bool inPositionOnFlop)
    {
        this->tableType = tt;
        this->position = position;
        this->numCallers = numCallers;
        this->numRaises = numRaises;
        this->numPlayers = numPlayers;
        this->previousAction = previousAction;
        this->inPositionOnFlop = inPositionOnFlop;
    }

    bool PreFlopParams::forcedAction() const
    {
        bool forced = true;

        if (position == Position_BigBlind)
        {
            forced = numRaises > 0;
        }

        return forced;
    }

    int PreFlopParams::toIndex() const
    {
        int index = 0;

        if (tableType == TableType_HU)
        {
            if (position == Position_SmallBlind)
            {
                index = std::min(numRaises, 4);
            }
            else if (position == Position_BigBlind)
            {
                index = 5 + std::min(numRaises, 4);
            }
            else
            {
                throw std::runtime_error("Bad position in heads-up game");
            }
        }
        else if (previousAction == Action_Fold) // Prvi krug
        {
            int a0 = (int)position;
            int a1 = -1;
                
            if (numRaises == 0)
            {
                a1 = (numCallers == 0) ? 0 : a1;
                a1 = (numCallers > 0) ? 1 : a1;
            }
            else if (numRaises == 1)
            {
                a1 = (numCallers == 0) ? 2 : a1;
                a1 = (numCallers > 0) ? 3 : a1;
            }
            else if (numRaises >= 2)
            {
                a1 = 4;
            }
 
            assert((a0 != -1) && (a1 != -1));
            index = (5 * a0) + a1; 
        }
        else // Drugi krug. PreviousAction: Check/Call, Raise/Bet
        {
            int a0 = -1;
            int a1 = -1;
            int a2 = -1;
            int a3 = -1;

            a0 = (previousAction == Action_Check || previousAction == Action_Call) ? 0 : a0;
            a0 = (previousAction == Action_Bet || previousAction == Action_Raise) ? 1 : a0;

            a1 = inPositionOnFlop ? 0 : 1;

            a2 = (numPlayers == 2) ? 0 : a2;
            a2 = (numPlayers > 2) ? 1 : a2;

            if (numRaises == 1)
            {
                a3 = (numCallers == 0) ? 0 : a3;
                a3 = (numCallers > 0) ? 1 : a3;
            }
            else if (numRaises >= 2)
            {
                a3 = 2;
            }

            assert((a0 != -1) && (a1 != -1) && (a2 != -1) && (a3 != -1));
            index = 30 + (12 * a0) + (6 * a1) + (3 * a2) + a3;
        }

        return index;
    }
}
